rust/lance-index/src/vector/hnsw.rs

```diff
@@ -17,21 +17,146 @@
+use arrow_array::{
+    builder::{ListBuilder, UInt32Builder},
+    cast::AsArray,
+    ListArray, RecordBatch, UInt32Array,
+};
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use lance_core::{Error, Result};
+use lance_file::{reader::FileReader, writer::FileWriter};
 use lance_linalg::distance::MetricType;
+use lance_table::io::manifest::ManifestDescribing;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use snafu::{location, Location};
 
-use super::graph::{Graph, OrderedFloat};
+use super::graph::{
+    builder::GraphBuilder, storage::VectorStorage, Graph, OrderedFloat, NEIGHBORS_COL,
+    NEIGHBORS_FIELD,
+};
 use crate::vector::graph::beam_search;
 pub mod builder;
 pub use builder::HNSWBuilder;
 
+const HNSW_TYPE: &str = "HNSW";
+const POINTER_COL: &str = "__pointer";
+
+lazy_static::lazy_static! {
+    /// POINTER field.
+    ///
+    pub static ref POINTER_FIELD: Field = Field::new(POINTER_COL, DataType::UInt32, true);
+}
+
+/// One level of the HNSW graph.
+///
+/// Type parameters:
+/// ----------------
+/// T : Float
+///   The data type of the vectors?
+struct HnswLevel {
+    /// All the nodes in this level.
+    // TODO: we just load the whole level into memory without pagation.
+    nodes: RecordBatch,
+
+    /// A reference to the neighbors of each node.
+    ///
+    /// Keep a reference of this array so that `neighbors()` can return reference
+    /// without lifetime issue.
+    neighbors: Arc<ListArray>,
+
+    /// The values of the neighbors array.
+    neighbors_values: Arc<UInt32Array>,
+
+    /// Vector storage of the graph.
+    vectors: Arc<dyn VectorStorage>,
+}
+
+impl HnswLevel {
+    /// Load one Hnsw level from the file.
+    async fn load(
+        reader: &FileReader,
+        row_range: Range<usize>,
+        vectors: Arc<dyn VectorStorage>,
+    ) -> Result<Self> {
+        let nodes = reader.read_range(row_range, reader.schema(), None).await?;
+        Ok(Self::new(nodes, vectors))
+    }
+
+    /// Create a new Hnsw Level from the nodes and the vector storage object.
+    fn new(nodes: RecordBatch, vectors: Arc<dyn VectorStorage>) -> Self {
+        let neighbors: Arc<ListArray> = nodes
+            .column_by_name(NEIGHBORS_COL)
+            .unwrap()
+            .as_list()
+            .clone()
+            .into();
+        let values: Arc<UInt32Array> = neighbors.values().as_primitive().clone().into();
+        Self {
+            nodes,
+            neighbors,
+            neighbors_values: values,
+            vectors,
+        }
+    }
+
+    fn from_builder(builder: &GraphBuilder, vectors: Arc<dyn VectorStorage>) -> Result<Self> {
+        let mut neighbours_builder = ListBuilder::new(UInt32Builder::new());
+        let mut pointers_builder = UInt32Builder::new();
+
+        for (_, node) in builder.nodes.iter() {
+            neighbours_builder.append_value(node.neighbors.values().map(|&n| Some(n)));
+            pointers_builder.append_value(node.pointer);
```

Comment:
What are pointers?  Row addresses?

Reply:
This is more a freedom to each graph implementation, it can be used as "id of a vector", or in HNSW case, this is "the id of the node on the next level"